docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md

```diff
@@ -0,0 +1,615 @@
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace FirstAnalyzerCS
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class MakeConstAnalyzer : DiagnosticAnalyzer
+    {
+        public const string DiagnosticId = "MakeConstCS";
+        private const string Title = "Variable can be made constant";
+        private const string MessageFormat = "Can be made constant";
+        private const string Description = "Make Constant";
+        private const string Category = "Usage";
+
+        private static DiagnosticDescriptor Rule = new DiagnosticDescriptor(DiagnosticId, Title, MessageFormat, Category, DiagnosticSeverity.Warning, isEnabledByDefault: true, description: Description);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get { return ImmutableArray.Create(Rule); } }
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterSyntaxNodeAction(AnalyzeNode, SyntaxKind.LocalDeclarationStatement);
+        }
+
+        private static void AnalyzeNode(SyntaxNodeAnalysisContext context)
+        {
+            throw new NotImplementedException();
+        }
+    }
+}
+```
+
+### Spot declarations that are already const
+
+Now you're ready to write the logic to determine whether a local variable can be declared as a const in the AnalyzeNode method. You'll make these changes in steps as you learn about the APIs available to write code that understands code.
+
+Start by performing some syntactic analyzer when your action is invoked. Open **MakeConst.cs**. In the `AnalyzeNode` method, cast the node passed in to the <xref:Microsoft.CodeAnalysis.CSharp.Syntax.LocalDeclarationStatementSyntax> type. You can safely assume this cast will succeed because the registered action specifies that type of node.
+
+```C#
+var localDeclaration = (LocalDeclarationStatementSyntax)context.Node;
+```
+
+To ensure the best performance, the general structure of an action method is to quickly eliminate cases where it won't create or report a diagnostic. The MakeConst analyzer can exit immediately if a local declaration is already a constant. To illustrate this construct, you'll write your first test.
+
+The structure of the unit tests is shown in the two example tests created by the template. Open **MakeConstUnitTests.cs**. The first test, `TestMethod1` shows the structure of a test when you want to verify that analyzed code does not trigger a diagnostic. `TestMethod2` shows the structure of a test when analyzed code triggers a diagnostic and provides a fix.
+
+Change `TestMethod1` to the following code:
+
+[!code-csharp[Don't report diagnostics on const declarations](~/samples/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst/MakeConst.Test/MakeConstUnitTests.cs#AlreadyConst "Test to ensure no diagnostics are reported on already const definition")]
```

Comment:
Pretty much all of the linked code samples have really bad indentation and look bad in the article.

Reply:
I want to make sure I make the right fix for this comment.  The code in magenta are string declarations that represent the source code being analyzed in a test. This is a typical test format for a diagnostic. See [roslyn-analyzer tests](https://github.com/dotnet/roslyn-analyzers/blob/master/src/Microsoft.CodeAnalysis.Analyzers/UnitTests/FixAnalyzers/FixerWithFixAllAnalyzerTests.Fixer.cs) for examples. If that's what you're concerned about, I'll highlight that better when it is first introduced.

Or do you mean the other code in the includes?

Reply:
No I mean in the preview article, all of the code samples pulled in are indented incorrectly.

Reply:
Thanks for walking in through this with me. 

I came up with a plan to make the sample tutorial more readable, and keep the unit test structure more clear. I'll update the samples this morning, and work on the article (addressing the other existing feedback) this afternoon.